### ISSUE TITLE
Getting Helm release history seems to read all secrets from the cluster

### ISSUE THREAD

**Issue by hifi** [rancher/fleet#1637]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

The original problem is/was that our etcd had something causing a lot of read traffic from it.

![image](https://github.com/rancher/fleet/assets/106598/85569027-6230-4511-b6af-a2e913fdd47b)

In this screenshot we have/had around 100MB/s going out from one etcd replica constantly. The drop in the graph is when I scaled down fleet-agent for a moment.

### Steps To Reproduce

Looking into the issue more, I manually created a bunch of big secrets on a dev cluster in some completely unrelated namespace with a script:

```sh
#!/bin/bash
kubectl create namespace secret-benchmark

for i in `seq 1 50`; do
    kubectl -n secret-benchmark apply --server-side -f - <<EOF
---
apiVersion: v1
kind: Secret
type: Opaque
metadata:
  name: blob-$i
data:
  blob: $(dd if=/dev/urandom bs=1M count=1 status=none | base64 -w0)
EOF
done
```

After creating these secrets I could again see the constant read traffic from etcd, with v0.6.0. Now, v0.6.0 seemed to get into a forever loop fetching a non-existent Helm release so it kept logging this:

```
time="2023-07-06T11:03:22Z" level=info msg="getting history for release development-XXXX-managed-system-agent"
```

These seem to match the read burst from etcd when I look at the raw increase of read bytes from etcd at the time of that log line.

Upgrading fleet-agent to v0.7.0 seemed to fix the loop, but it still did a huge burst read on startup when it got the history for all releases once. This is good, but it doesn't seem to be fixed.

### Environment

```markdown
- Architecture:amd64
- Fleet Version: v0.6.0/v0.7.0
- Cluster:
  - Provider: RKE
  - Options:
  - Kubernetes Version: v1.25.9+rke2r1
```

### Anything else?

Looking into the code I _suspect_ that the issue might be around here:

https://github.com/rancher/fleet/blob/6729c95b2b94910b42d49b85e9e8eb01fb8a553c/internal/helmdeployer/deployer.go#L116

The history getter later on uses `globalCfg` without `getCfg()` to limit it to a namespace so it might cause that.

https://github.com/rancher/fleet/blob/6729c95b2b94910b42d49b85e9e8eb01fb8a553c/internal/helmdeployer/deployer.go#L598-L599

Probably not, but just something I noticed.

Also #1609 seems related.

**Comment by moio** [comment]
This is interesting @hifi, thanks for posting.

According to my understanding, `hist.Run` boils down to

https://github.com/helm/helm/blob/293b50c65d4d56187cd4e2f390f0ada46b4c4737/pkg/storage/storage.go#L156

which further delegates to

https://github.com/helm/helm/blob/v3.11.1/pkg/storage/driver/secrets.go#L112-L123

which calls `List` without specifying a namespace but actually specifying a label selector, which should contain the helm-only `owner=helm` label and the release specific `name=<releasename>` label.

Would it be possible for you to retrieve the Kubernetes API audit log, to double check if my reasoning above is actually correct?

**Comment by hifi** [comment]
That actually seems to be the reason in this case. Since there's no namespace selector the API will pull all Secrets just to test if they have those labels.

I ran this fairly simple test case on my local machine that won't even return any results but the impact is around the same (with the exception of kubectl htting all API servers between requests):

```sh
$ while true; do kubectl get secret --all-namespaces -l foo=bar; sleep 1; done
```

![image](https://github.com/rancher/fleet/assets/106598/cd02dfdf-295d-4c54-a4b0-d763adce7c5e)

**Comment by hifi** [comment]
Meanwhile we upgraded Rancher to v2.7.5 and it had the expected effects on another production cluster:

![image](https://github.com/rancher/fleet/assets/106598/e2e2dcd2-4dcc-4732-9c92-2973890f9042)

Now fleet-agent (v0.7.0) doesn't go fetching those Helm releases in a loop anymore so etcd traffic settled down but the underlying issue of Helm pulling all Secrets from all namespaces during that label selector filter will always hit etcd and API hard when its run on our clusters.

Not sure if there's any downstream work that can be done outside not listing releases unless absolutely needed to.

**Comment by moio** [comment]
The fleet team is aware of the issue and one of the options being considered is actually alternative Helm storage designs - eg. an ad-hoc Helm `Driver` implementation.

We hope to have a finalized decision, and hopefully fixes, coming out soon.

cc @manno for awareness

**Comment by weyfonk** [comment]
See also: #1809.
TL;DR: Fleet now uses a new Helm storage driver with a secrets cache, meaning that even if unnecessary retrieval over all namespaces still happens (which will be addressed in a future PR), this should no longer result in as high traffic to/from etcd.

**Comment by manno** [comment]
This is now tracked in #1812
